### PR TITLE
Adds external authentication through vapor-auth

### PR DESCRIPTION
**Pull request by ghost**
:cl:
server: Added external authentication, so you can login even when the BYOND hub is down! Just use the "Register/Update-Account" verb in the OOC tab!
/:cl:

you can login just using a password prompt, or by using the special fancy smhancy vapor-auth-client for autologin.

uses https://github.com/steamp0rt/vapor-auth (both client and server are in that repo)

i could've implemented this via a server-side threading dll but this is easier and less likely to lag the server

### CODE REVIEW

File: hippiestation/code/__HELPERS/unsorted.dm [hunk 1]
```diff
@@ -7,3 +7,53 @@
...
+			if(len-i > 1)
+				buffer |= text2ascii(str, i+2)
+				. += ascii2text(text2ascii(_base64AlphaStr, 1 + (buffer >> 6))) + \
+					ascii2text(text2ascii(_base64AlphaStr, 1 + (buffer & 0x3F)))
+			else . += ascii2text(text2ascii(_base64AlphaStr, 1 + (buffer >> 6)))// + "="
+		else . += ascii2text(text2ascii(_base64AlphaStr, 1 + (buffer >> 6))) + \
+			ascii2text(text2ascii(_base64AlphaStr, 1 + (buffer & 0x3F)))// + \
+			//"=="
+
+/proc/base64bin(string)
```
Comment (yoyobatty) [inline hippiestation/code/__HELPERS/unsorted.dm:31]: That's some math right there kids.

Comment (ghost) [reply hippiestation/code/__HELPERS/unsorted.dm:31]: both of those base64 procs are a decade old, minimum. lol.

**Comment by ghost** [comment]
Ya know, I just realized, DLL + /world/Topic based auth is still an option. If ya'll would actually prefer this method (there'd be no visible difference for players -- but you wouldn't have to host a seperate webserver/API, at the cost of possible bits of lag on the server itself due to DLL calls), tell me

**Comment by ghost** [comment]
Some important notes:

- `auth.sql` is needed in the DB for both methods. Even if you use vapor-auth-server, the game still needs to track which ckey = what key (because ckey = canonical key), and also people's last login location (for autofilling usernames)

- If you use the DLL method, its needs to be compiled in PostCompile.sh

**Comment by yoyobatty** [comment]
Fix gonflicts

File: hippiestation/code/modules/authentication/unauthed.dm [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+		client.player_details.byond_version = "[client.byond_version].[client.byond_build || "xxx"]"
+		GLOB.player_details[new_key] = client.player_details
+	client.key = new_key
+	GLOB.directory[C.ckey] = C
+	var/list/banned = world.IsBanned(C.ckey, C.address, C.computer_id, real_bans_only=TRUE)
+	if(LAZYLEN(banned))
+		overlay_fullscreen("ban_message", /obj/screen/fullscreen/ban_message)
+		to_chat(C, "<span class='boldannounce big'>You are banned from the server.</span>")
+		to_chat(C, banned["desc"])
+		sleep(5)
```
Comment (yoyobatty) [inline hippiestation/code/modules/authentication/unauthed.dm:64]: REE?

File: hippiestation/code/modules/authentication/unauthed.dm [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+		qdel(C)
+		qdel(src)
+		return
+	C.InitClient(tdata)
+	if(istype(C.mob, /mob/dead/unauthed))
+		C.mob = new /mob/dead/new_player
```
Comment (yoyobatty) [inline hippiestation/code/modules/authentication/unauthed.dm:70]: C.mob = new /mob/dead/new_player(src)

Should that not be the way to do it? Or do we move the player out nullspace afterwards?

Comment (ghost) [reply hippiestation/code/modules/authentication/unauthed.dm:70]: They get force moved out by Initialize anyways

File: hippiestation/code/modules/authentication/unauthed.dm (hunk 2 above)
Comment (ghost) [reply hippiestation/code/modules/authentication/unauthed.dm:64]: It's only half-a-second, and it's just to ensure the ban message shows.

Also, not spammable.

File: hippiestation/code/modules/authentication/unauthed.dm (hunk 3 above)
Comment (ghost) [reply hippiestation/code/modules/authentication/unauthed.dm:70]: https://github.com/HippieStation/HippieStation/blob/master/code/modules/mob/dead/new_player/new_player.dm#L25-L28

**Comment by Manly-Dwarf** [comment]
this still a thing?

**Comment by ghost** [comment]
@steamp0rt what's preventing this from being merged?

**Comment by ghost** [comment]
@HideAndSeekLOGIC this is something that kinda needs the OK from Gregg himself

**Comment by AnthonyJNgo** [comment]
+1 for the idea, but implementation needs to be drawn out more so that we can see if it is feasible or not.

I'm going to close this for now because it is stale, but if you pick this project back up just make a new PR.

**Comment by ghost** [comment]
> I'm going to close this for now because it is stale, but if you pick this project back up just make a new PR.

it's stale because the code's pretty much finished

**Pull request closed without merging**